File: src/python/Trees/PreOrderTraversal.py
class TreeNode:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

def preorder_traversal(root):
    if root:
        print(root.value, end=' ')  # Process the root
        preorder_traversal(root.left)  # Traverse left subtree
        preorder_traversal(root.right)  # Traverse right subtree

if __name__ == '__main__':
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)

    preorder_traversal(root)

File: src/java/Trees/PreOrderTraversal.java
import Trees.TreeNode;

public class PreOrderTraversal{
  public void preorderTraversal(TreeNode root) {
        if (root == null) {
            return;
        }
        System.out.print(root.val + " ");  // Process the root
        preorderTraversal(root.left);  // Traverse left subtree
        preorderTraversal(root.right);  // Traverse right subtree
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);

        PreOrderTraversal tree = new PreOrderTraversal();
        tree.preorderTraversal(root);
    }
}
